tests/system/test_asyncpg_direct_connection.py

```diff
@@ -0,0 +1,152 @@
+            time = conn.execute(sqlalchemy.text("SELECT NOW()")).fetchone()
+            conn.commit()
+
+    Args:
+        ip_address (str):
+            The IP address of an AlloyDB instance, e.g., 10.0.0.1
+        user (str):
+            The formatted IAM database username.
+            e.g., my-email@test.com, service-account@project-id.iam
+        db_name (str):
+            The name of the database, e.g., mydb
+    """
+    # [START alloydb_sqlalchemy_asyncpg_connect_iam_authn_direct]
+    engine = create_async_engine(
+        # Equivalent URL:
+        # postgresql+asyncpg://<user>:empty@<host>:5432/<db_name>
+        sqlalchemy.engine.url.URL.create(
+            drivername="postgresql+asyncpg",
+            username=user,  # IAM db user, e.g. service-account@project-id.iam
+            password="",  # placeholder to be replaced with OAuth2 token
+            host=ip_address,  # AlloyDB instance IP address
+            port=5432,
+            database=db_name,  # "my-database-name"
+        ),
+        connect_args={"ssl": "require"},
+    )
+
+    # set 'do_connect' event listener to replace password with OAuth2 token
+    # must use engine.sync_engine as async events are not implemented
+    @event.listens_for(engine.sync_engine, "do_connect")
+    def auto_iam_authentication(dialect, conn_rec, cargs, cparams) -> None:
+        cparams["password"] = get_authentication_token(creds)
+
+    # [END alloydb_asyncpg_connect_iam_authn_direct]
+    return engine
+
+
+async def test_sqlalchemy_asyncpg_time() -> None:
+    """Basic test to get time from database using asyncpg with SQLAlchemy."""
+    ip_address = os.environ["ALLOYDB_INSTANCE_IP"]  # Private IP for AlloyDB instance
+    user = os.environ["ALLOYDB_IAM_USER"]
+    db = os.environ["ALLOYDB_DB"]
+
+    engine = create_sqlalchemy_engine(ip_address, user, db)
+    # [START alloydb_sqlalchemy_asyncpg_connect_iam_authn_direct]
+    # use connection from connection pool to query AlloyDB database
+    async with engine.connect() as conn:
+        result = await conn.execute(sqlalchemy.text("SELECT NOW()"))
+        time = result.fetchone()
+        print("Current time is ", time[0])
+        # [END alloydb_sqlalchemy_asyncpg_connect_iam_authn_direct]
+        curr_time = time[0]
+        assert type(curr_time) is datetime
+    # cleanup AsyncEngine
+    await engine.dispose()
+
+
+async def test_native_asyncpg_time() -> None:
+    """Basic test to get time from database using native asyncpg connection."""
+    ip_address = os.environ["ALLOYDB_INSTANCE_IP"]  # Private IP for AlloyDB instance
+    user = os.environ["ALLOYDB_IAM_USER"]
+    db = os.environ["ALLOYDB_DB"]
+
+    # [START alloydb_native_asyncpg_connect_iam_authn_direct]
+    # ... inside of async context (function)
+    async with asyncpg.create_pool(
+        user=user,  # IAM db user, e.g. service-account@project-id.iam
+        password=get_authentication_token(creds),  # set OAuth2 token as password
```

Comment:
Will get_authentication_token be a `callable` here? This looks like a one-off call to my eyes, rather than a callable.